include/rosparam_shortcuts/rosparam_shortcuts.h

```diff
@@ -106,6 +110,246 @@ bool convertDoublesToEigen(const std::vector<double>& values, Eigen::Isometry3d&
+
+	mutable std::mutex mutex_;  // protect parameters
+	mutable std::vector<std::function<void()>> changed_notify_callback_;  // enable user to have callbacks for changed
+	                                                                      // parameters
+
+	rclcpp::node_interfaces::OnSetParametersCallbackHandle::SharedPtr on_set_callback_handle_;  // ros2 callback method
+	                                                                                            // for updating
+	                                                                                            // parameters
+};
+
+NodeParameters::NodeParameters(const rclcpp::Node::SharedPtr& node, const rclcpp::Logger& logger)
+  : node_(node), logger_(logger) {}
+
+template <typename ParameterT>
+void NodeParameters::declare_and_get(const std::string& name, const ParameterT& default_value,
+                                     const rcl_interfaces::msg::ParameterDescriptor& parameter_descriptor,
+                                     const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name, default_value, parameter_descriptor);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+template <typename ParameterT>
+void NodeParameters::declare_and_get(const std::string& name, const ParameterT& default_value,
+                                     const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name, default_value);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+template <typename ParameterT>
+void NodeParameters::declare_and_get(const std::string& name, const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+void NodeParameters::register_ros_callback_reconfigure() {
+	on_set_callback_handle_ = node_->add_on_set_parameters_callback(
+	    [this](const std::vector<rclcpp::Parameter>& parameters) { return set_parameters_callback(parameters); });
+}
+
+void NodeParameters::register_update_callback(const std::function<void()>& callback) const {
+	changed_notify_callback_.push_back(callback);
+}
+
+bool NodeParameters::get(const std::string& name, rclcpp::Parameter& parameter) const {
+	auto internal_param = parameters_.find(name);
```

Comment:
the get function should probably lock the mutex so it isn't reading from the maps while they could be modified by the callback